docs/architecture/adr-050-sign-mode-textual.md

```diff
@@ -129,6 +130,44 @@ Msg (<int>/<int>): <string>           // E.g. Msg (1/2): bank v1beta1 send coins
 End of transaction messages
 ```
 
+### 9. Custom `Msg` Renderers
+
+Application developers may choose to not follow default renderer value output for their own `Msg`s. In this case, they can implement their own custom `Msg` renderer. This is similar to [EIP4430](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4430.md), where the smart contract developer chooses the description string to be shown to the end user.
+
+This is done by setting the `cosmos.msg.v1.textual.custom_renderer` Protobuf option to a non-empty string. This option CAN ONLY be set on a Protobuf message representing transaction message object (implementing `sdk.Msg` interface).
+
+```proto
+message MsgFooBar {
+  // Optional comments to describe in human-readable language the formatting
+  // rules of the custom renderer.
+  option (cosmos.msg.v1.textual.custom_renderer) = "<unique algorithm identifier>";
+
+  // proto fields
+}
+```
+
+When this option is set on a `Msg`, a registered function will transform the `Msg` into an array of one or more strings, which MAY use the key/value format (described in point #3) with the expert field prefix (described in point #5). These strings MAY be rendered from a `Msg` field using a default value renderer, or they may be generated from several fields using custom logic.
+
+The `unique algorithm identifier` is a string convention chosen by the application developer and is used to identify the custom `Msg` renderer. For example, the SPEC documentation of this custom algorithm can reference this identifier. This identifier CAN have a versioned suffix (e.g. `_v1`) to adapt for future changes (which would be consensus-breaking). We also recommend adding Protobuf comments to describe in human language the custom logic used.
+
+Moreover, the renderer must provide 2 functions: one for formatting from Protobuf to string, and one for parsing string to Protobuf. These 2 functions are provided by the application developer. To satisfy point #1, these 2 functions MUST be bijective with each other. Bijectivity of these 2 functions will not be checked by the SDK at runtime. However, we strongly recommend the application developer to include a comprehensive suite in their app repo to test bijectivity, as to not introduce security bugs. A simple bijectivity test looks like:
```

Comment:
A bijective rendering between two representations of a type is more accurately called an encoding. Do you actually intend that Cosmos should define, implementing, and maintain its own bespoke Protobuf encoding scheme?

Reply:
Yes. Hence the discussion [here](https://github.com/cosmos/cosmos-sdk/pull/12785#discussion_r939445681) about naming this custom encoding of Msgs "expert" or "unsafe".

Reply:
Does that mean implementing this encoding for a given msg type is optional?

Reply:
Yes. By default, all Msgs are encoded with the default value renderers (see annex 1). If the app developer wishes to overwrite the default encoding, then it uses this custom Msg renderer.

Reply:
Ah, okay! If a msg doesn't implement this textual/expert/unsafe renderer (encoder) will it still be sign-able on Ledger devices?

Reply:
1. Custom renderers are completely optional - Msgs with no custom renderer registered will use the default message rendering described in the annex.
2. I'd argue to include an option for runtime verification of bijectionality when custom extensions are used, but otherwise, there's a possibility that the signed text omits some important information. However, the risks of this must be weighed against the risks of illegibility from non-customized or non-textual presentations.
3. The Msg to text rendering for a given concrete Msg must be the same on the user's client software and on the validators, so the part that dispatches from the Msg, proto, or field value to the renderer must be deterministic.

Reply:
> I'd argue to include an option for runtime verification of bijectionality when custom extensions are used, but otherwise, there's a possibility that the signed text omits some important information. However, the risks of this must be weighed against the risks of illegibility from non-customized or non-textual presentations.> The Msg to text rendering for a given concrete Msg must be the same on the user's client software and on the validators, so the part that dispatches from the Msg, proto, or field value to the renderer must be deterministic.



```
type MyMsg struct {
    Name string
    Desc string
}

func (m *MyMsg) RenderJSON() ([]byte, error) { return json.Marshal(m) }
func (m *MyMsg) RenderText() ([]byte, error) { return m.Name }
```

Reply:
> The Msg to text rendering for a given concrete Msg must be the same on the user's client software and on the validators, so the part that dispatches from the Msg, proto, or field value to the renderer must be deterministic.

I'm not worried about the behavior of a given renderer, I'm worried about the notion that the signable bytes for a given tx/msg aren't well-defined but are instead produced by one of N arbitrary methods on the type. As an example, given

```
type MyMsg struct {
    Name string
    Desc string
}

func (m *MyMsg) RenderJSON() ([]byte, error) { return json.Marshal(m) }
func (m *MyMsg) RenderText() ([]byte, error) { return []byte(m.Name + "\n" + m.Desc) }
```

Is it the case that the signature for this msg could be based on either of these two methods depending on a parameter that's not part of the msg? If so, hopefully this example demonstrates why this design is unsound? 😬

Reply:
There will need to be some method of registering a particular custom renderer and telling the dispatcher which message type it's supposed to handle. For instance, in Go you'd have something like the following in `app.go`:

```
textual := valuerenderer.NewTextual()
textual.RegisterCustom(NewMyMsgTextRenderer(), MyMsg{})
tx.EnableTextual(textual)
```
 
The exact details would be part of the implementation, not necessarily part of the spec, since there will be both Go and Javascript implementations of this functionality, each will have its own appropriate API.

Reply:
If signer S signs msg M, is that signature always the same, or does it depend on the "renderer" of that msg?

Reply:
The signature S of a Msg M depends on the signing mode (Textual, Direct, Legacy Amino, Legacy Amino JSON) chosen for the transaction, and even more basically it depends on the particular blockchain in question, since some Msg types will be supported on some blockchains and not on others. So for a particular revision of software for a particular blockchain, and a particular signing mode for a particular message, then yes, that must determine a sequence of bytes to feed into the signing algorithm for signature and for verification.

The feature discussed in this PR allows a particular revision of a particular blockchain to customize the default SDK behavior when using the Textual mode, but once it has done so, it must be deterministic for a given Msg.

Reply:
Ah, so a given transaction on a given blockchain at a given height will always have precisely one signing mode? Great!

Reply:
No - given all the preceding *and* the signing mode, it will have precisely representation that is input to signing. For a given blockchain and txn you can choose to sign either the text mode translation, or the JSON mode translation, or whatever.

Reply:
So that's a big problem :) 

A tx/msg is Merkelized as a sequence of bytes stored on chain. If you sign anything other than that specific sequence of bytes, then you're not signing the tx/msg itself, you're signing a transformation. The transform function isn't part of consensus, so you can't assume sign_mode_textual will always do the same thing. And it's basically not possible to assert that a transformation of one byte sequence to another is equivalent and lossless at a semantic (i.e. the protobuf) level, so you can't assume that verified signatures in different modes have actually signed the same information and establish equivalent trust.

IMO this is a critical, SEV-0, red-alert kind of problem. But take that feedback as you like 😉

Reply:
Your feedback seems like it applies to Sign Mode Textual as a whole, not the specific changes contemplated in this PR, so I'll ask you to take your concerns to a more appropriate venue. Perhaps Discord.